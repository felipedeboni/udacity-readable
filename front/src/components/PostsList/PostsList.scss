.post__list__container {
  .form-inline {
    text-align: center;

    @media( min-width: 767px ) {
      text-align: right;
    }

    .control-label {
      margin-bottom: 0;
      vertical-align: middle;
    }

    .form-control {
      display: inline-block;
      width: auto;
      vertical-align: middle;
    }
  }
}

.post__list__not__found {
  color: #999;
  font-size: 18px;
  padding: 10vh 0;
}
