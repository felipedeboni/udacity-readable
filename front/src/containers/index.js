export { default as AppContainer } from './AppContainer/AppContainer';

export {
  default as CommentContainer
} from './CommentContainer/CommentContainer';
export {
  default as CommentFormContainer
} from './CommentFormContainer/CommentFormContainer';
export {
  default as CommentsListContainer
} from './CommentsListContainer/CommentsListContainer';
export {
  default as CommentScoreContainer
} from './CommentScoreContainer/CommentScoreContainer';

export { default as PostContainer } from './PostContainer/PostContainer';
export {
  default as PostFormContainer
} from './PostFormContainer/PostFormContainer';
export {
  default as PostsListContainer
} from './PostsListContainer/PostsListContainer';
export {
  default as PostScoreContainer
} from './PostScoreContainer/PostScoreContainer';
