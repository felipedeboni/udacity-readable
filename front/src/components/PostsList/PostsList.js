import React, { Component } from 'react';
import PropTypes from 'prop-types';
import sortBy from 'sort-by';
import { PostContainer } from 'containers';
import { PostListItem } from 'components';
import './PostsList.css';

class PostsList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      sort: props.sort || '-timestamp'
    };
  }

  handleChange = value => {
    this.setState({
      sort: value
    });
  };

  render() {
    const { posts } = this.props;
    const { sort } = this.state;
    const sortedPosts = posts.sort(sortBy(sort));

    return (
      <section className="post__list__container">
        {sortedPosts.length > 0 && (
          <div className="form-inline">
            <label htmlFor="sortby" className="control-label">
              Sort By
            </label>&nbsp;
            <select
              id="sortby"
              onChange={e => this.handleChange(e.target.value)}
              defaultValue={sort}
              className="form-control"
            >
              <option value="-timestamp">Newest</option>
              <option value="timestamp">Oldest</option>
              <option value="-voteScore">Highest Score</option>
              <option value="voteScore">Lowest Score</option>
            </select>
          </div>
        )}
        <ul className="post__list list-unstyled">
          {sortedPosts.length > 0 &&
            sortedPosts.map(post => (
              <PostContainer
                key={post.id}
                post={post}
                component={PostListItem}
              />
            ))}
          {sortedPosts.length === 0 && (
            <div className="post__list__not__found text-center">
              Whoops! Looks like we don't have posts yet :(
            </div>
          )}
        </ul>
      </section>
    );
  }
}

PostsList.propTypes = {
  posts: PropTypes.array.isRequired,
  sort: PropTypes.oneOf(['-timestamp', 'timestamp', '-voteScore', 'voteScore'])
};

export default PostsList;
