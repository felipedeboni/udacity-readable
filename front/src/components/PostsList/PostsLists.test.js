import React from 'react';
import { shallow } from 'enzyme';
import PostsList from './PostsList';
import { PostContainer } from 'containers';

const { posts } = global.mock;

describe('PostList', () => {
  it('renders without crashing', () => {
    const test = <PostsList posts={posts} sort="-timestamp" />;

    expect(shallow(test)).toMatchSnapshot();
  });

  it('renders with empty message', () => {
    const test = <PostsList posts={[]} />;

    expect(shallow(test)).toMatchSnapshot();
  });

  it('should handle change', () => {
    const test = shallow(<PostsList posts={posts} sort="-timestamp" />);
    const instance = test.instance();
    const handleChange = jest.fn();
    instance.handleChange = handleChange;
    test.find('.form-inline select').simulate('change', {
      target: {
        value: 'voteScore'
      }
    });
    expect(handleChange).toHaveBeenCalled();
  });

  it('should sort', () => {
    const test = shallow(<PostsList posts={posts} sort="-timestamp" />);
    test.find('.form-inline select').simulate('change', {
      target: {
        value: '-voteScore'
      }
    });
    test.update();

    const firstScore = test
      .find(PostContainer)
      .first()
      .props().post.voteScore;
    const lastScore = test
      .find(PostContainer)
      .last()
      .props().post.voteScore;
    expect(firstScore > lastScore).toBe(true);
  });
});
