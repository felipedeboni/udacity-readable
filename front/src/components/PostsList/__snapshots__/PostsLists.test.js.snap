// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList renders with empty message 1`] = `
<section
  className="post__list__container"
>
  <ul
    className="post__list list-unstyled"
  >
    <div
      className="post__list__not__found text-center"
    >
      Whoops! Looks like we don't have posts yet :(
    </div>
  </ul>
</section>
`;

exports[`PostList renders without crashing 1`] = `
<section
  className="post__list__container"
>
  <div
    className="form-inline"
  >
    <label
      className="control-label"
      htmlFor="sortby"
    >
      Sort By
    </label>
    Â 
    <select
      className="form-control"
      defaultValue="-timestamp"
      id="sortby"
      onChange={[Function]}
    >
      <option
        value="-timestamp"
      >
        Newest
      </option>
      <option
        value="timestamp"
      >
        Oldest
      </option>
      <option
        value="-voteScore"
      >
        Highest Score
      </option>
      <option
        value="voteScore"
      >
        Lowest Score
      </option>
    </select>
  </div>
  <ul
    className="post__list list-unstyled"
  >
    <Connect(PostContainer)
      component={[Function]}
      key="6ni6ok3ym7mf1p33lnez"
      post={
        Object {
          "author": "thingone",
          "body": "Just kidding. It takes more than 10 minutes to learn technology.",
          "category": "redux",
          "commentCount": 0,
          "deleted": false,
          "id": "6ni6ok3ym7mf1p33lnez",
          "timestamp": 1468479767190,
          "title": "Learn Redux in 10 minutes!",
          "voteScore": -5,
        }
      }
    />
    <Connect(PostContainer)
      component={[Function]}
      key="8xf0y6ziyjabvozdd253nd"
      post={
        Object {
          "author": "thingtwo",
          "body": "Everyone says so after all.",
          "category": "react",
          "commentCount": 2,
          "deleted": false,
          "id": "8xf0y6ziyjabvozdd253nd",
          "timestamp": 1467166872634,
          "title": "Udacity is the best place to learn React",
          "voteScore": 6,
        }
      }
    />
  </ul>
</section>
`;
